Readme.md
# node-smartcash
[![npm](https://img.shields.io/badge/npm-1.1.2-blue.svg)](https://npmjs.org/package/node-smartcash)

[smartcash 1.2.0+](https://github.com/SmartCash/Core-Smart/releases)

node-smartcash is based on [node-bitcoin](https://github.com/freewil/node-bitcoin) with modifications to disable SSL, to allow an array of command parameters, and to update the list of API commands (based on the [official list of SmartCash console commands](https://smartcash.freshdesk.com/support/solutions/articles/35000027144-debug-console-command-line)). You can either use the CamelCase method on the `smartcash.Client` object (methods are listed in [lib/commands.js](/lib/commands.js)) or use the `cmd` method directly. More information about JSON-RPC usage for SmartCash core can be found in the [SmartCash config file reference](https://smartcash.freshdesk.com/support/solutions/articles/35000038702-smartcash-conf-configuration-file) (search "JSON-RPC options"). More information about a particular command can be found via Help > Debug Window > Console in SmartCash Core (SmartCash node client) by using `help [command]` where [command] is the command in lowercase.

node-smartcash is best used as a dependency in a project that will communicate with a SmartCash Core instance on the same machine. For an example see [SmartSweeper](https://github.com/swiftlettech/smart-sweeper).

## Install

`npm install node-smartcash`

## Examples

`const smartcash = require('node-smartcash')`

### Create client
```js
// all config options are optional
var client = new smartcash.Client({
  host: 'localhost',
  port: 9678,
  user: 'username',
  pass: 'password',
  timeout: 30000
});
```

### Get the balance across all accounts with a minimum number of 6 confirmations for all transactions

```js
client.getBalance(['*', 6], function(error, response, resHeaders) {
  if (error) console.log('Error: ', error);
  console.log('Balance: ', response);
});
```
### Getting the balance directly using `cmd`

```js
client.cmd('getbalance', ['*', 6], function(error, response, resHeaders){
  if (error) console.log('Error: ', error);
  console.log('Balance: ', response);
});
```

### Get information about a command

```js
client.help(['getbalance'], function(error, response, resHeaders) {
  if (error) console.log('Error: ', error);
  console.log(response);
});
```
### Getting information about a command directly using `cmd`

```js
client.cmd('help', ['getbalance'], function(error, response, resHeaders){
  if (error) console.log('Error: ', error);
  console.log(response);
});
```

## SSL

Unlike the original node-bitcoin connecting to a remote bitcoind instance, using JSON-RPC over SSL to connect to a remote SmartCash Core instance isn't supported.


## Tests

`npm test`

To run the tests you must have [SmartCash Core installed](https://smartcash.cc/wallets/) and configured to accept RPC commands. The rpcuser and rpcpassword in the SmartCash config file (smartcash.conf) must match user and pass in [test/config.js](test/config.js). See the [SmartCash config file reference](https://smartcash.freshdesk.com/support/solutions/articles/35000038702-smartcash-conf-configuration-file) (search "JSON-RPC options") for more information.


Changelog.md
# node-smartcash changelog

## v1.1.2 (2019/03/24)
* Added getRandomKeyPair

## v1.1.1 (2018/06/25)
* Added debug, dumpHdInfo, getAddressBalance, getAddressDeltas, getAddressMempool, getAddressTxids, getAddressUtxos, getBlockHashes, getBlockHeaders, getGenerate, getSpentInfo, importElectrumWallet, instantSendToAddress, setGenerate, setNetworkActive, smartMining, smartRewards, spork (added in smartcash-1.2.0)
* Removed addWitnessAddress, generateToAddress, getMempoolAncestors, getMempoolDescendants, getMempoolEntry, getPoolInfo, importPrunedFunds, removePrunedFunds, signMessageWithPrivKey (removed in smartcash-1.2.0)

## v1.1.0 (2018/05/28)
* Removed SSL support completely (not supported in SmartCash core)
* Updated lib/commands.js
* Updated tests

## v1.0.0 (2018/05/21)
* Changed license from [Unlicense](https://choosealicense.com/licenses/unlicense/) to [MIT](https://choosealicense.com/licenses/mit/)
* Updated the callRpc function in lib/jsonrpc.js to allow arrays of command parameters
* Updated lib/commands.js to include missing SmartCash commands and remove bitcoin-only commands
* Removed information about sending a batch of RPC calls from the readme (not supported in SmartCash core)
* Removed information about JSON-RPC over SSL from the readme (not recommended)

# node-bitcoin changelog

## v3.0.1 (2015/10/25)
* fix redefinition of already defined variable (does not actually affect behavior)

## v3.0.0 (2015/10/18)
* make public domain license explicit
* remove commands dropped in bitcoind v0.11
  * `getHashesPerSecond`
  * `getHashesPerSec`
* add missing commands for bitcoind v0.11
  * `generate`
  * `verifyTxOutProof`

## v2.4.0 (2015/07/16)
* don't lazy-load http/https modules
* add command for bitcoind v0.11.0: `getTxOutProof`

## v2.3.2 (2015/06/26)
* fix bug in test suite that was supposed to detect missing commands
* add missing commands
  * `prioritiseTransaction`
  * `importWallet`

## v2.3.1 (2015/06/24)
* add missing `getMempoolInfo` for bitcoind v0.10

## v2.3.0 (2015/02/04)
* drop node v0.8.x support
* update testnet-box
* update devDependencies
* add commands for bitcoind v0.10
  * `estimateFee`
  * `estimatePriority`
  * `getChainTips`
  * `importAddress`

## v2.2.0 (2014/08/29)
* add commands for bitcoind v0.9.x
  * `decodeScript`
  * `dumpWallet`
  * `getBestBlockHash`
  * `getBlockchainInfo`
  * `getNetTotals`
  * `getNetworkInfo`
  * `getNetworkHashPs`
  * `getRawChangeAddress`
  * `getUnconfirmedBalance`
  * `getWalletInfo`
  * `ping`
  * `verifyChain`

## v2.1.2 (2014/04/16)
* lazy load `http`/`https` module

## v2.1.1 (2014/03/25)
* change default request timeout from `5000`ms to `30000`ms

## v2.1.0 (2014/03/12)
* remove `deprecate` dependency
* add request timeout option (defaults to 5000ms)
* add 3rd parameter to callbacks: response headers

## v2.0.1 (2014/01/08)
* default `host` to 'localhost'; `port` to '8332'

## v2.0.0 (2013/10/14)
* remove deprecated commands
  * `getMemoryPool`
  * `getMemorypool`
* remove deprecated functionality
  * creating `bitcoin.Client` with more than one argument

## v1.7.0 (2013/05/05)
* add missing commands from bitcoind v0.7.0
  * `createMultiSig`
  * `getBlockTemplate`
  * `getTxOut`
  * `getTxOutSetInfo`
  * `listAddressGroupings`
  * `submitBlock`
* deprecate commands
  * `getMemoryPool`
  * `getMemorypool`

## v1.6.2 (2013/03/21)
* shrink package size via .npmignore

## v1.6.1 (2013/03/13)
* add node v0.10.x support (rejectUnauthorized defaults to true in 0.10.x)

## v1.6.0 (2013/03/08)
* drop node v0.6.x support
* change test runner from `vows` to `mocha`
* upgrade testnet-box
* add commands for bitcoind v0.8.0
  * `addNode`
  * `getAddedNodeInfo`
  * `listLockUnspent`
  * `lockUnspent`
* deprecate creating `bitcoin.Client` with more than one argument
* add SSL support

## v1.5.0 (2012/10/22)
* remove `getBlockNumber` test
* upgrade testnet-box
* add RPC call batching (multiple RPC calls within one HTTP request)

## v1.4.0 (2012/09/09)
* add commands for bitcoind v0.7.0
  * `createRawTransaction`
  * `decodeRawTransaction`
  * `getPeerInfo`
  * `getRawMemPool`
  * `getRawTransaction`
  * `listUnspent`
  * `sendRawTransaction`
  * `signRawTransaction`
* remove deprecated `getBlockNumber`

## v1.3.1 (2012/08/19)
Remove `underscore` dependency

## v1.3.0 (2012/07/03)
Change use of http.createClient() (deprecated in node v0.8.x) to http.request()

## v1.2.2 (2012/04/26)
Fix callback being called twice when a client and request error
occur on the same command call.

## v1.2.1 (2012/04/26)
* add missing `getBlock` command

## v1.2.0 (2012/04/25)
* submodule testnet-box for running tests
* err objects should all now be an instance of Error

## v1.1.6 (2012/04/11)
* add commands for bitcoind v0.6.0
  * `addMultiSigAddress` (only available in testnet)
  * `dumpPrivKey`
  * `getBlockHash`
  * `getMiningInfo`
  * `importPrivKey`


test/test.js
/* global describe, it */

var assert = require('assert')
var clone = require('clone')
var http = require('http')
var smartcash = require('../')
var config = require('./config')

var test = {
  account: 'Original'
}

var makeClient = function makeClient() {
    return new smartcash.Client(config)
}

var notEmpty = function notEmpty(data) {
    if (data === 0) return
    assert.ok(data)
}

describe('General Client Tests', function () {
    it('smartcash related error should be an Error object', function (done) {
    var client = makeClient()
    client.cmd('nomethod', function (err, expectedValue) {
      assert.ok(err instanceof Error)
      assert.equal(err.message, 'Method not found')
      assert.equal(err.code, -32601)
      assert.equal(expectedValue, undefined)
      done()
    })
  })

  describe('help()', function () {
    it('should return help', function (done) {
      var client = makeClient()
      client.help(function (err, help) {
        assert.ifError(err)
        notEmpty(help)
        done()
      })
    })
  })
    
  describe('invalid credentials', function () {
    var badCredentials = clone(config)
    badCredentials.user = 'baduser'
    badCredentials.pass = 'badpwd'
    var client = new smartcash.Client(badCredentials)

    it('should still return client object', function (done) {
      assert.ok(client instanceof smartcash.Client)
      done()
    })

    it('should return status 401 with html', function (done) {
      client.getDifficulty(function (err, difficulty) {
        assert.ok(err instanceof Error)
        assert.equal(err.message, 'Invalid params, response status code: 401')
        assert.equal(err.code, -32602)
        assert.equal(difficulty, undefined)
        done()
      })
    })
  })

  describe('creating client on non-listening port', function () {
    var badPort = clone(config)
    badPort.port = 9897
    badPort.user = 'baduser'
    badPort.pass = 'badpwd'
    var client = new smartcash.Client(badPort)

    it('will return client object', function (done) {
      assert.ok(client instanceof smartcash.Client)
      done()
    })

    it('should not call callback more than once', function (done) {
      client.listSinceBlock(function (err, result) {
        assert.ok(err instanceof Error)
        done()
      })
    })
  })

  describe('request timeouts', function () {
    it.skip('should occur by default after 30000ms', function (done) {
      this.timeout(31000)
      var request // eslint-disable-line no-unused-vars
      var client = new smartcash.Client({
        host: 'localhost',
        port: 19998,
        user: 'admin1',
        pass: '123'
      })
      var start = Date.now()
      client.getInfo(function (err, info) {
          console.log(err)
          
        var delta = Date.now() - start
        assert.ok(err instanceof Error)
        assert.ok(err.code === 'ETIMEDOUT' || err.code === 'ESOCKETTIMEDOUT')
        assert.ok(delta >= 30000, 'delta should be >= 30000: ' + delta)
        done()
      })
    })

    it.skip('should be customizable', function (done) {
      this.timeout(4500)
      var request // eslint-disable-line no-unused-vars
      var client = new smartcash.Client({
        host: 'localhost',
        port: 19999,
        user: 'admin1',
        pass: '123',
        timeout: 2500
      })
      var start = Date.now()
      client.getInfo(function (err, info) {
        var delta = Date.now() - start
        assert.ok(err instanceof Error)
        assert.ok(err.code === 'ETIMEDOUT' || err.code === 'ESOCKETTIMEDOUT')
        assert.ok(delta >= 2500, 'delta should be >= 2500:' + delta)
        done()
      })
    })
  })
})

describe('Client Blockchain Tests', function () {
  describe('getDifficulty()', function () {
    it('should get difficulty', function (done) {
      var client = makeClient()
      client.getDifficulty(function (err, difficulty) {
        assert.ifError(err)
        assert.ok(typeof difficulty === 'number')
        done()
      })
    })
  })
    
  describe('getInfo()', function () {
    it('should get info', function (done) {
      var client = makeClient()
      client.getInfo(function (err, info, headers) {
        assert.ifError(err)
        notEmpty(info)
        assert.ok(info.errors === '')
        done()
      })
    })
  })
    
  describe('response headers', function () {
    var assertResHeaders = function (resHeaders) {
      assert.ok(resHeaders)
      assert.ok(resHeaders['content-type'])
      assert.ok(resHeaders.date)
      assert.ok(resHeaders['content-length'])
      assert.ok(resHeaders.connection)
    }
    it('should be returned for no parameter calls', function (done) {
      var client = makeClient()
      client.getInfo(function (err, info, resHeaders) {
        assert.ifError(err)
        notEmpty(info)
        assert.ok(info.errors === '')
        assertResHeaders(resHeaders)
        done()
      })
    })
  })
})

describe('Client Wallet Tests', function () {
  describe('getAccountAddress()', function () {
    it('should be able to get an account address', function (done) {
      var client = makeClient()
      client.getAccountAddress([test.account], function (err, address) {
        assert.ifError(err)
        assert.ok(address)
        client.getAccount([address], function (err, account) {
          assert.ifError(err)
          assert.equal(account, test.account)
          done()
        })
      })
    })
  })
    
  describe('getBalance()', function () {
    it('should return balance without any args', function (done) {
      var client = makeClient()
      client.getBalance(function (err, balance) {
        assert.ifError(err)
        assert.ok(typeof balance === 'number')
        done()
      })
    })
  })
    
  describe('listTransactions()', function () {
    it('should be able to listTransactions with specific count', function (done) {
      var client = makeClient()
      client.listTransactions([test.account, 15], function (err, txs) {
        assert.ifError(err)
        assert.ok(txs)
        assert.ok(Array.isArray(txs))
        done()
      })
    })

    it('should be able to listTransactions without specific count', function (done) {
      var client = makeClient()
      client.listTransactions([test.account], function (err, txs) {
        assert.ifError(err)
        assert.ok(txs)
        assert.ok(Array.isArray(txs))
        done()
      })
    })
  })
    
  describe('response headers', function () {
    var assertResHeaders = function (resHeaders) {
      assert.ok(resHeaders)
      assert.ok(resHeaders['content-type'])
      assert.ok(resHeaders.date)
      assert.ok(resHeaders['content-length'])
      assert.ok(resHeaders.connection)
    }
    it('should be returned for 1-parameter call', function (done) {
      var client = makeClient()
      client.getNewAddress([test.account], function (err, address, resHeaders) {
        assert.ifError(err)
        assert.ok(address)
        assertResHeaders(resHeaders)
        done()
      })

      it('should be returned for 2-parameter call', function (done) {
        var client = makeClient()
        client.listTransactions([test.account, 15], function (err, txs, resHeaders) {
          assert.ifError(err)
          assert.ok(txs)
          assert.ok(Array.isArray(txs))
          assertResHeaders(resHeaders)
          done()
        })
      })
    })
  })
})

test/config.js
module.exports = {
  host: 'localhost',
  port: 9678,
  user: 'rpcusername',
  pass: 'rpcpassword',
  timeout: 30000
};


test/missing-commands.js
/* global describe, it */

var assert = require('assert')
var smartcash = require('../')
var config = require('./config')
var commands = require('../lib/commands')

var getHelpCommands = function (client, cb) {
  var commandRegex = /^([a-z]+)/
  client.cmd('help', function (err, commandList) {
    if (err) return cb(err)

    var helpCommands = []

    // split up the command list by newlines
    var commandListLines = commandList.split('\n')
    var result
    for (var i in commandListLines) {
      result = commandRegex.exec(commandListLines[i])
      if (!result) continue
      helpCommands.push(result[1])
    }
    cb(null, helpCommands)
  })
}

describe('Client Commands', function () {
  it('should have all the commands listed by `help`', function (done) {
    var client = new smartcash.Client(config)
    getHelpCommands(client, function (err, helpCommands) {
      assert.ifError(err)

      for (var i in helpCommands) {
        var found = false
        for (var j in commands) {
          if (commands[j] === helpCommands[i]) {
            found = true
            break
          }
        }
          
        // ignore walletlock, walletpassphrase, and walletpassphrasechange in help because they are hidden
        // if the wallet isn't encrypted and ignore encryptwallet because it's hidden if the wallet is encrypted
        var ignore = ['encryptwallet', 'walletlock', 'walletpassphrase', 'walletpassphrasechange']
        if (~ignore.indexOf(helpCommands[i]))
          assert.ok(!found, 'missing command not found in `help`: ' + helpCommands[i])
        else
          assert.ok(found, 'missing command found in `help`: ' + helpCommands[i])
      }

      done()
    })
  })

  it('should not have any commands not listed by `help`', function (done) {
    var client = new smartcash.Client(config)
    getHelpCommands(client, function (err, helpCommands) {
      assert.ifError(err)

      for (var i in commands) {
        var found = false
        for (var j in helpCommands) {
          if (commands[i] === helpCommands[j]) {
            found = true
            break
          }
        }

        // ignore walletlock, walletpassphrase, and walletpassphrasechange in help because they are hidden
        // if the wallet isn't encrypted and ignore encryptwallet because it's hidden if the wallet is encrypted
        var ignore = ['encryptwallet', 'walletlock', 'walletpassphrase', 'walletpassphrasechange']
        if (~ignore.indexOf(commands[i])) {
          assert.ok(!found, 'command found in `help`: ' + commands[i])
        } else {
          assert.ok(found, 'command not found in `help`: ' + commands[i])
        }
      }

      done()
    })
  })
})


lib/commands.js
module.exports = {
  abandonTransaction: 'abandontransaction',
  addMultiSigAddress: 'addmultisigaddress',
  addNode: 'addnode',
  backupWallet: 'backupwallet',
  clearBanned: 'clearbanned',
  createMultiSig: 'createmultisig',
  createRawTransaction: 'createrawtransaction',
  debug: 'debug',
  decodeRawTransaction: 'decoderawtransaction',
  decodeScript: 'decodescript',
  disconnectNode: 'disconnectnode',
  dumpHdInfo: 'dumphdinfo',
  dumpPrivKey: 'dumpprivkey',
  dumpWallet: 'dumpwallet',
  estimateFee: 'estimatefee',
  estimatePriority: 'estimatepriority',
  estimateSmartFee: 'estimatesmartfee',
  estimateSmartPriority: 'estimatesmartpriority',
  fundRawTransaction: 'fundrawtransaction',
  generate: 'generate',
  getAccount: 'getaccount',
  getAccountAddress: 'getaccountaddress',
  getAddedNodeInfo: 'getaddednodeinfo',
  getAddressBalance: 'getaddressbalance',
  getAddressDeltas: 'getaddressdeltas',
  getAddressMempool: 'getaddressmempool',
  getAddressesByAccount: 'getaddressesbyaccount',
  getAddressTxids: 'getaddresstxids',
  getAddressUtxos: 'getaddressutxos',
  getBalance: 'getbalance',
  getBestBlockHash: 'getbestblockhash',
  getBlock: 'getblock',
  getBlockchainInfo: 'getblockchaininfo',
  getBlockCount: 'getblockcount',
  getBlockHash: 'getblockhash',
  getBlockHashes: 'getblockhashes',
  getBlockHeader: 'getblockheader',
  getBlockHeaders: 'getblockheaders',
  getBlockTemplate: 'getblocktemplate',
  getChainTips: 'getchaintips',
  getConnectionCount: 'getconnectioncount',
  getDifficulty: 'getdifficulty',
  getGenerate: 'getgenerate',
  getInfo: 'getinfo',
  getMempoolInfo: 'getmempoolinfo',
  getMiningInfo: 'getmininginfo',
  getNetTotals: 'getnettotals',
  getNetworkHashPs: 'getnetworkhashps',
  getNetworkInfo: 'getnetworkinfo',
  getNewAddress: 'getnewaddress',
  getPeerInfo: 'getpeerinfo',
  getRandomKeyPair: 'getrandomkeypair',
  getRawChangeAddress: 'getrawchangeaddress',
  getRawMemPool: 'getrawmempool',
  getRawTransaction: 'getrawtransaction',
  getReceivedByAccount: 'getreceivedbyaccount',
  getReceivedByAddress: 'getreceivedbyaddress',
  getSpentInfo: 'getspentinfo',
  getTransaction: 'gettransaction',
  getTxOut: 'gettxout',
  getTxOutProof: 'gettxoutproof',
  getTxOutSetInfo: 'gettxoutsetinfo',
  getUnconfirmedBalance: 'getunconfirmedbalance',
  getWalletInfo: 'getwalletinfo',
  help: 'help',
  importAddress: 'importaddress',
  importElectrumWallet: 'importelectrumwallet',
  importPrivKey: 'importprivkey',
  importPubKey: 'importpubkey',
  importWallet: 'importwallet',
  instantSendToAddress: 'instantsendtoaddress',
  keypoolRefill: 'keypoolrefill',
  listAccounts: 'listaccounts',
  listAddressGroupings: 'listaddressgroupings',
  listBanned: 'listbanned',
  listLockUnspent: 'listlockunspent',
  listReceivedByAccount: 'listreceivedbyaccount',
  listReceivedByAddress: 'listreceivedbyaddress',
  listSinceBlock: 'listsinceblock',
  listTransactions: 'listtransactions',
  listUnspent: 'listunspent',
  lockUnspent: 'lockunspent',
  move: 'move',
  ping: 'ping',
  prioritiseTransaction: 'prioritisetransaction',
  sendFrom: 'sendfrom',
  sendMany: 'sendmany',
  sendRawTransaction: 'sendrawtransaction',
  sendToAddress: 'sendtoaddress',
  setAccount: 'setaccount',
  setBan: 'setban',
  setGenerate: 'setgenerate',
  setNetworkActive: 'setnetworkactive',
  setTxFee: 'settxfee',
  signMessage: 'signmessage',
  signRawTransaction: 'signrawtransaction',
  smartMining: 'smartmining',
  smartnode: 'smartnode',
  smartnodeBroadcast: 'smartnodebroadcast',
  smartnodeList: 'smartnodelist',
  smartRewards: 'smartrewards',
  snSync: 'snsync',
  spork: 'spork',
  stop: 'stop',
  submitBlock: 'submitblock',
  validateAddress: 'validateaddress',
  verifyChain: 'verifychain',
  verifyMessage: 'verifymessage',
  verifyTxOutProof: 'verifytxoutproof'
}

lib/jsonrpc.js
var http = require('http')

var Client = function (opts) {
  this.opts = opts || {}
  this.http = http
}

Client.prototype.call = function (method, params, callback, errback, path) {
  var time = Date.now()
  var requestJSON

  if (Array.isArray(method)) {
    // multiple rpc batch call
    requestJSON = []
    method.forEach(function (batchCall, i) {
      requestJSON.push({
        id: time + '-' + i,
        method: batchCall.method,
        params: batchCall.params
      })
    })
  } else {
    // single rpc call
    requestJSON = {
      id: time,
      method: method,
      params: params
    }
  }

  // First we encode the request into JSON
  requestJSON = JSON.stringify(requestJSON)

  // prepare request options
  var requestOptions = {
    host: this.opts.host || 'localhost',
    port: this.opts.port || 9678,
    method: 'POST',
    path: path || '/',
    headers: {
      'Host': this.opts.host || 'localhost',
      'Content-Length': requestJSON.length
    },
    agent: false
  }

  // use HTTP auth if user and password set
  if (this.opts.user && this.opts.pass) {
    requestOptions.auth = this.opts.user + ':' + this.opts.pass
  }

  // Now we'll make a request to the server
  var cbCalled = false
  var request = this.http.request(requestOptions)

  // start request timeout timer
  var reqTimeout = setTimeout(function () {
    if (cbCalled) return
    cbCalled = true
    request.abort()
    var err = new Error('ETIMEDOUT')
    err.code = 'ETIMEDOUT'
    errback(err)
  }, this.opts.timeout || 30000)

  // set additional timeout on socket in case of remote freeze after sending headers
  request.setTimeout(this.opts.timeout || 30000, function () {
    if (cbCalled) return
    cbCalled = true
    request.abort()
    var err = new Error('ESOCKETTIMEDOUT')
    err.code = 'ESOCKETTIMEDOUT'
    errback(err)
  })

  request.on('error', function (err) {
    if (cbCalled) return
    cbCalled = true
    clearTimeout(reqTimeout)
    errback(err)
  })

  request.on('response', function (response) {
    clearTimeout(reqTimeout)

    // We need to buffer the response chunks in a nonblocking way.
    var buffer = ''
    response.on('data', function (chunk) {
      buffer = buffer + chunk
    })
    // When all the responses are finished, we decode the JSON and
    // depending on whether it's got a result or an error, we call
    // emitSuccess or emitError on the promise.
    response.on('end', function () {
      var err

      if (cbCalled) return
      cbCalled = true

      try {
        var decoded = JSON.parse(buffer)
      } catch (e) {
        if (response.statusCode !== 200) {
          err = new Error('Invalid params, response status code: ' + response.statusCode)
          err.code = -32602
          errback(err)
        } else {
          err = new Error('Problem parsing JSON response from server')
          err.code = -32603
          errback(err)
        }
        return
      }

      if (!Array.isArray(decoded)) {
        decoded = [decoded]
      }

      // iterate over each response, normally there will be just one
      // unless a batch rpc call response is being processed
      decoded.forEach(function (decodedResponse, i) {
        if (decodedResponse.hasOwnProperty('error') && decodedResponse.error != null) {
          if (errback) {
            err = new Error(decodedResponse.error.message || '')
            if (decodedResponse.error.code) {
              err.code = decodedResponse.error.code
            }
            errback(err)
          }
        } else if (decodedResponse.hasOwnProperty('result')) {
          if (callback) {
            callback(decodedResponse.result, response.headers)
          }
        } else {
          if (errback) {
            err = new Error(decodedResponse.error.message || '')
            if (decodedResponse.error.code) {
              err.code = decodedResponse.error.code
            }
            errback(err)
          }
        }
      })
    })
  })
  request.end(requestJSON)
}

module.exports.Client = Client
